### ISSUE TITLE
Discussion: Features for Realism Overhaul's mods

### ISSUE THREAD

**Issue by Dunbaratu** [KSP-KOS/KOS#2075]
(Note: the "Not Ready" label, which is normally for Pull Requests, is on this Issue to represent that the issue *itself* isn't even done being typed up yet.  I'm adding more to it as I go over the next few weeks.)

While I'm not sure if this belongs in an Addons module, or if it's even do-able that way, I'm experimenting with Realism Overhaul to discover what things kOS doesn't expose for RO that make RO and kOS hard to use together.

### This should be split into more than one issue before starting to work on it.

I'm using this as a single checklist for discussion and brainstorming.  When actually implementing things from this list, it would be best to split them off into their own issues.

### Which mod does what?

One complication is that RO isn't one mod.  It's a mod PACK, which means each of these features really comes from some other mod, which could be installed independently without the rest of RO present.  Therefore the trigger conditions for whether or not to add these features would have to be per mod inside RO, rather than RO itself.  (i.e. instead of "If RO installed, then.....", we'd have to make it "If RealFuels installed, then...." and "If Deadly Re-Entry installed then...", etc for each mod that matters to the bullet points below.)  For each point, we'd have to identify the actual mod that causes the feature to exist, and test for THAT mod (rather than RO) to enable the feature.  On a first pass, I haven't even identified which mod does each of these things.  Having help with that would be tremendous.

**This is a checklist I plan to add more to later.**

For each item in the list, I will mention if I found a workaround that lets you get the info some other (but maybe more complex) way.  Things for which there is no workaround should have higher priority.

I categorize workarounds into these three types:

- Yes : There is a workaround that's only somewhat cumbersome.
- Yes But : There is a workaround but it's ugly and may not be practical.
- No : There is no workaround at all that I could think of.

(For this list, the name "**RCM**" will be used as an abbreviation for the "Right Click Menu" of a part).

### Engine part suffixes:
[ ] *Current Ullage state*:  Ullage is the requirement that makes engines fail if the fuel is floating around in freefall.  Ullage state ("Very stable", "unstable", "risky", etc....) is communicated to the user via a string on the RCM.
    - Workaround Exists? Yes - you can access this as a ``KSPField`` on the RCM.
[ ] *Ullage Required*: Some engines never worry about Ullage.  Others do.  This should be communicated to the user as a Boolean.  (A script's behaviour when dealing with an engine that can never have Ullage problems can differ from when dealing with an engine that can.)
    - Workaround Exists? Yes But - The user could maintain their own homemade database of engine info.
[ ] *minthrust*: Many engines will flameout below a min threshold setting.  Not only does a script need to know this minimum so it can calculate burns, but it doesn't even know what a throttle setting *means* without this information.  (The RO game scales the input throttle range of 0..1 into the allowed range for the engine.  A throttle setting of 0.0001 doesn't mean "slightly above zero" anymore.  Now it means "slightly above the engine's flameout minimum".  A setting of 0.5 doesn't mean "half of max" anymore.  Now it means "halfway between min and max".)  We could either communicate this as a coefficient, (i.e. 0.7 to mean an engine that can't go below 70% of maxthrust) or as just a minthrust as an absolute (i.e. in kilonewtons just like maxthrust is.)  The name for the suffix should match which choice we make.  (Only call it "minthrust" if it's in kilonewtons.  Call it something else if it's a coefficient number).
    - Workaround Exists? Yes But - The user could maintain their own homemade database of engine info.
[ ] *ignitions left*: The engines often have limited number of restarts, and after that you can't ignite them.  Note that setting the throttle to 0 counts as using up an ignition.  (it takes a new ignition to start up the engine again).  This number should be available to scripts.
    - Workaround Exists? Yes - you can access this as a ``KSPField`` on the RCM.

### kUniverse-style stuff

[ ] Build time: The Kerbal Build Time mod introduces a delay between when you try to launch a vehicle design and when it arrives at the launchpad ready to go.  You can't just hit "launch" and expect the script to be able to fly the vessel at once that instant.  It may be a good idea to be able to read these delay times if a script is trying to use kUniverse to auto-launch missions.  (There is also a chance that our kUniverse features will allow a player to cheat around this enforced delay by accessing the launching API directly instead of through the UI.  Not sure if there's anything we can do about that or would even want to.  If they want to cheat, let them.)

### Avionics limits and kos for all

[ ] Ability to remove KOSProcessor from a part while building in the VAB.  RO uses the "kos for all" mod, which inserts a kOS computer unit into *every single* part that is identified as some kind of probe core or command module.  While this may make sense from a certain point of view, it stops making sense when you realize that RO's requirement for "avionics support" mandates that you have several "extra" probe cores all over the rocket that exist purely for avionics.  These shouldn't all have to have kOS computers in them.  Doing so probably introduces quite a bit of lag, and even if it doesn't it still uglifies the UI by making you have to sift through a long list of kOS terminal buttons to find the right one.  Since "kos for all" is merely a ModuleManager config, and all it does is just add the KOSProcessor PartModule to every core, it can't really differentiate between probe cores that merely exist for avionics boost and ones that are meant to be controlling cores.  There's a UI slider in the VAB to change this, but (As far as I can tell) it doesn't seem to do anything.  It may help for us to add an option to the RCM in the VAB/SPH that will allow a part with a KOSProcessor in it to have its KOSProcessor PartModule removed.

**Comment by Dunbaratu** [comment]
I plan to expand this list as I come up with more things.  This is very much a work in progress.

**Comment by Patrykz94** [comment]
This is kind of separate from RO, but maybe getting the engine status from TestFlight mod?
Would be useful to know if the engine is working nominaly or if there are any failures like, thrust loss, performance loss, or a shutdown.
I guess rated burntime and elapsed burntime (not sure if that's a correct term?) could be useful as well in some situations.

**Comment by Dunbaratu** [comment]
@Patrykz94 : TestFlight is *sort of* part of RO, in that it's one of RO's sideline recommended but not strictly required mods.  I didn't use it because I like KSP's original design principle that all failures be deterministically the player's fault, but I can see the argument for including it as part of RO's suite of things.  But as someone who doesn't much use it, it may be low on my list (as it's something that takes a lot of going out of my way to learn about it and test it.)

Is the information you mention something that exists on the rightclick menu for the part?  If so, there is a workaround that's a bit ugly but can be done (find out which PartModule causes that label to appear on the part's rightclick menu, and you can write a kerboscript function that will read the value.)  If not, then it rises to a higher priority in my eyes because there's nothing a script can do to read it without new C# code in the kOS mod itself.

**Comment by hvacengi** [comment]
Due to the loose connection to RO, I would recommend this integration being made into a separate issue

**Comment by hvacengi** [comment]
>In a scene load, RT connections don't exist for a few ticks. RT seems to need a few physics ticks after loading the scene before noticing a relay connection exists. So kOS boot scripts, even if they wait until ship:unpacked, still have no connection to the archive to load files from, even when they should, unless they wait a while. Maybe we could have a part of the RT addon that you can query to find out if RT is "ready"??
>    Workaround Exists? Yes - all boot scripts that try to copy from archive need to wait a second or so at the top.

Instead of waiting a few seconds, wouldn't a good workaround be `wait until homeconnection:isconnected`?  At which point, I actually think this is a better choice anyways, since you shouldn't assume that you have a connection when the ship loads.  If that code doesn't work,, we definitely need to fix that.

**Comment by Dunbaratu** [comment]
>wait until homeconnection:isconnected?

If there is no connection there still may be other things you want the boot script to do.  (a common use pattern in a boot script is "if there's an updated version, copy it.  If not, use the old version I already have locally").  There is no good way for a script to query "is `homeconection:isconnected` capable of returning true yet or not?"  If it is now *capable of* returning true yet because RT has had enough ticks to finish "waking up", and yet despite being capable of returning a True, it still returns a False anyway, then that False is a *real* False and you shouldn't expect it to turn true again just by waiting a few ticks.

Maybe we can find a way to find out from RT whether or not RT is ready yet after a load?  Then we could make homeconnection:isconnected into a blocking operation that yeilds until RT says it's ready and we know the answer it returns is trustworthy.

**Comment by SlavDimov** [comment]
I would like to suggest integration with the TestFlight mod.
I am going mad when TestFlight cuts the performance to half or shuts down one of the engines and my whole craft gets unbalanced and crashes. I tried reading the engine values and saw that, for engines that are unthrottlable the mod reduces the fuel flow to make them perform worse and I guess I can use that to detect when an engine is broken, but for engines that are throttlable the fuel flow gets reduced when the throttle is reduced, and not linearly so I cant use that to detect if engine is broken.
I looked over the GitHub page of the mod and saw an API folder. Perhaps the KOS itself can tie into it in order to provide information about the engine status?

Other than that the engine statuses are show in the mod's gui from the stock toolbar, but I don't know if it is possible to read that.

Thanks.
